### PR TITLE
Fix for #1248: Hosting with BindCommandLine throws TargetParameterCountException

### LINKED ISSUE #1248: Hosting with BindCommandLine throws TargetParameterCountException
I am trying to move from System.CommandLine.Hosting 0.3.0-alpha.20371.2 to 0.3.0-alpha.20574.7
the code which runs in previous version without an issue, throws suddenly an System.Reflection.TargetParameterCountException: 'Parameter count mismatch.'.

I created a small repro to nail the issue a little bit down.
I think the issue comes from #1032 

**Sample Code:**
```csharp
    internal class ExampleClass
    {
        public string Sample { get; set; }
    }

    internal class MyCommandOptions
    {
        //public string Path { get; set; }
        //public int Number { get; set; }
        //public SampleEnum EnumValue { get; set; } = SampleEnum.One;
        public List<ExampleClass> List { get; set; }
    }

    internal class Program
    {
        static async Task Main(string[] args) => await BuildCommandLine()
            .UseDefaults()
            .UseHost(_ => Host.CreateDefaultBuilder(),
                host =>
                {
                    host.ConfigureServices((ctx, services) =>
                                           {
                                               services.AddOptions();
                                               services.AddOptions<MyCommandOptions>()
                                                       //.Bind(ctx.Configuration)
                                                       .BindCommandLine();
                                           });
                })
            .Build()
            .InvokeAsync(args);

        private static CommandLineBuilder BuildCommandLine()
        {
            var root = new RootCommand(@"$ dotnet run --name 'Joe'"){
                //new Option<string>("--path"),
                //new Option<int>("--number"),
                //new Option<SampleEnum>("--EnumValue")
            };
            root.Handler = CommandHandler.Create<IHost>(Run);
            return new CommandLineBuilder(root);
        }

        private static void Run(IHost host)
        {
            var serviceProvider = host.Services;
            var options = serviceProvider.GetRequiredService<IOptions<MyCommandOptions>>();
            var settings = options.Value; // accessing the value throws exception
           //              ^^^^^^^^^^^^^
        }
```

I have a small options class `MyCommandOptions` which has some properties, in my real code it gets most of the configuration from the binding of a config json, but some it gets from the commandline binding.
For simplicity i commented the lines out, just to show the real example, but for the error they are irrelevant.

In my Code when i want to access the `options.Value` I now get the 'Parameter count mismatch.' Exception for the `public List<ExampleClass> List` which is not even added in Options.

**Call Stack:**
```
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimePropertyInfo.SetValue(Object obj, Object value, BindingFlags invokeAttr, Binder binder, Object[] index, CultureInfo culture)
   at System.Reflection.RuntimePropertyInfo.SetValue(Object obj, Object value, Object[] index)
   at System.Reflection.PropertyInfo.SetValue(Object obj, Object value)
   at System.CommandLine.Binding.PropertyDescriptor.SetValue(Object instance, Object value) in C:\Projects\command-line-api\src\System.CommandLine\Binding\PropertyDescriptor.cs:line 34
   at System.CommandLine.Binding.ModelBinder.UpdateInstanceInternalNotifyIfNonDefaultsUsed[T](T instance, BindingContext bindingContext) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 161
   at System.CommandLine.Binding.ModelBinder.InstanceFromSpecificConstructor(BindingContext bindingContext, ConstructorDescriptor constructor, IReadOnlyList`1 boundValues, Boolean& nonDefaultsUsed) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 141
   at System.CommandLine.Binding.ModelBinder.CreateInstanceInternal(BindingContext bindingContext) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 80
   at System.CommandLine.Binding.ModelBinder.GetBoundValue(IValueSource valueSource, BindingContext bindingContext, IValueDescriptor valueDescriptor, Boolean includeMissingValues, Type parentType) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 278
   at System.CommandLine.Binding.ModelBinder.GetBoundValues(IDictionary`2 bindingSources, BindingContext bindingContext, IReadOnlyList`1 valueDescriptors, Boolean enforceExplicitBinding, Type parentType, Boolean includeMissingValues) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 214
   at System.CommandLine.Binding.ModelBinder.UpdateInstanceInternalNotifyIfNonDefaultsUsed[T](T instance, BindingContext bindingContext) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 151
   at System.CommandLine.Binding.ModelBinder.UpdateInstance[T](T instance, BindingContext bindingContext) in C:\Projects\command-line-api\src\System.CommandLine\Binding\ModelBinder.cs:line 147
   at System.CommandLine.Hosting.HostingExtensions.<>c__4`1.<BindCommandLine>b__4_0(TOptions opts, IServiceProvider serviceProvider) in C:\Projects\command-line-api\src\System.CommandLine.Hosting\HostingExtensions.cs:line 81
   at Microsoft.Extensions.Options.ConfigureNamedOptions`2.Configure(String name, TOptions options)
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at Microsoft.Extensions.Options.OptionsManager`1.<>c__DisplayClass5_0.<Get>b__0()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
```

### PR DESCRIPTION
**Pull request by BobSilent**
Fixes  #1248

### CODE REVIEW

**Comment by jonsequitur** [comment]
Could you add a test?

**Comment by BobSilent** [comment]
updated with an explicit test case

File: src/System.CommandLine.Tests/Binding/ModelBinderTests.cs [hunk 1]
```diff
@@ -265,6 +265,23 @@ public void Parse_result_can_be_used_to_modify_an_existing_instance_without_doin
...
+            var instance = new ClassWithComplexTypes();
+            var bindingContext = new BindingContext(parser.Parse("the-command"));
+            var binder = new ModelBinder(typeof(ClassWithComplexTypes));
+
+            binder.UpdateInstance(instance, bindingContext);
+
+            instance.BoolOption.Should().Be(default);
+            instance.IntOption.Should().Be(default);
+            instance.ListOptionDefaultNull.Should().BeNull();
+            instance.ListOptionDefaultEmpty.Should().BeEmpty();
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/Binding/ModelBinderTests.cs:282]: You can simplify the assertions and make the intent a little clearer by doing this:

```cs
instance.Should().BeEquivalentTo(new ClassWithComplexTypes());
```

**Review by jonsequitur — APPROVED** [review]

**Comment by jonsequitur** [comment]
Thanks, @BobSilent!

**Pull request merged**